README.md

|  Version  | Docs  | Status |
| :--- 	  | :--- 	  | :--- 	  |
|  [![PyPi Version](https://img.shields.io/pypi/v/pynastran.svg)](https://pypi.python.org/pypi/pyNastran) | [![Documentation Status](https://readthedocs.org/projects/pynastran-git/badge/?version=v1.0)](http://pynastran-git.readthedocs.io/en/v1.0.0/?badge=v1.0.0) | [![Build Status](https://img.shields.io/travis/SteveDoyle2/pyNastran/v1.0.svg)](https://travis-ci.org/SteveDoyle2/pyNastran) [![Coverage Status](https://img.shields.io/coveralls/SteveDoyle2/pyNastran/v1.0.svg)](https://coveralls.io/github/SteveDoyle2/pyNastran?branch=v1.0) |
|   Master | [![Documentation Status](https://readthedocs.org/projects/pynastran-git/badge/?version=latest)](http://pynastran-git.readthedocs.io/en/latest/?badge=latest) | [![Linux Status](https://img.shields.io/travis/SteveDoyle2/pyNastran/master.svg)](https://travis-ci.org/SteveDoyle2/pyNastran) [![Windows Status](https://ci.appveyor.com/api/projects/status/1qau107h43mbgghi/branch/master?svg=true)](https://ci.appveyor.com/project/SteveDoyle2/pynastran) ![Coverage Status](https://coveralls.io/repos/github/SteveDoyle2/pyNastran/badge.svg?branch=master) | 



<!---
[![codecov](https://codecov.io/gh/SteveDoyle2/pyNastran/branch/master/graph/badge.svg)](https://codecov.io/gh/SteveDoyle2/pyNastran) 

[![Coverage Status](https://img.shields.io/coveralls/SteveDoyle2/pyNastran/master.svg)](https://coveralls.io/github/SteveDoyle2/pyNastran?branch=master)
--->

<!---

[![Build status](https://ci.appveyor.com/api/projects/status/1qau107h43mbgghi/branch/master?svg=true)](https://ci.appveyor.com/project/SteveDoyle2/pynastran/branch/master)

[![Build status](https://ci.appveyor.com/api/projects/status/1qau107h43mbgghi?svg=true)](https://ci.appveyor.com/project/SteveDoyle2/pynastran)

[![Build Status](https://travis-ci.org/SteveDoyle2/pyNastran.png)](https://travis-ci.org/SteveDoyle2/pyNastran)
[![Coverage Status](https://coveralls.io/repos/github/SteveDoyle2/pyNastran/badge.svg?branch=master)](https://coveralls.io/github/SteveDoyle2/pyNastran?branch=master)
[![Requirements Status](https://img.shields.io/requires/github/SteveDoyle2/pyNastran/master.svg)](https://requires.io/github/SteveDoyle2/pyNastran/requirements/?branch=master)	|
--->

### v1.0.0

[Download GUI](https://sourceforge.net/projects/pynastran/files/?source=navbar) (latest is from 2017/5/25)

<!---
[Doumentation](http://pynastran-git.readthedocs.org/en/latest/index.html)
--->

Also, check out the:
  * [Discussion forum](http://groups.google.com/group/pynastran-discuss) (intended for questions about the latest release)
  * [Developer forum](http://groups.google.com/group/pynastran-dev) (intended for questions about the master branch)
  * [Docs](http://pynastran-git.readthedocs.io/en/latest/?badge=latest) (the version isn't quite right, but it's close)

for more detailed information.

<!--- this isn't setup... -->
<!--- http://stevedoyle2.github.io/pyNastran/ --->

# Overview

pyNastran is an interface library to the various Nastran file formats (BDF, OP2, OP4).  Using the BDF interface, you can read/edit/write Nastran geometry without worrying about field formatting.  Many checks are also performed to verify that your model is correct.  Using the OP2 interface, you can read very large result files very quckly and very efficiently.  Additionally, you can also extract a subset of the result data and write F06 result files.

Using the pyNastran GUI, you can read in Nastran models and quickly view results for a model.  While it's no FEMAP/Patran, it can replace many tasks that would otherwise require a commercial program.

![GUI](https://github.com/SteveDoyle2/pynastran/blob/master/pyNastran/gui/images/caero.png)


<!--- Ripped off meshio  --->
<!--- [![Build Status](https://travis-ci.org/SteveDoyle2/pyNastran.svg?branch=master)](https://travis-ci.org/SteveDoyle2/pyNastran)  --->
<!--- [![codecov.io](https://codecov.io/github/SteveDoyle2/pyNastran/coverage.svg?branch=master)](https://codecov.io/github/SteveDoyle2/pyNastran?branch=master)  --->

<!--- ## pyNastran v0.8.0 has NOT been released (8/21/2016)   --->
<!--- [Download pyNastran v0.8] (https://github.com/SteveDoyle2/pyNastran/releases)  --->

<!--- 

This should be hidden...

### pyNastran v1.1.0 has not been released (6/25/2018)

It's been roughly a year and 100 tickets closed since the last version, so it's probably
time for another release!  I'm pleased to announce that the continuing problems 
of up-to-date documentation will hopefully be a thing of the past.  xxx 
has offered to host the documentation.  The pyNastranGUI exe (previously 
hosted on Sourceforge) will also be hosted there.  Outside of that, it's the same 
open-source project and will still be on Github.

Regarding features, the focus has again been on robustness and testing.  There has been
a 10% increase in the testing coverage (the same as v0.8 to v1.0).

There are a few changes (mainly in the BDF) though.  The GUI now also supports
PyQt4, PyQt5, and Pyside with the same API, which helps reduces the restriction on
licensing.

Programmatics
 - Dropping Python 3.4 support
 - dropping VTK 5/6 suppoprt for the GUI

BDF:
 - 343 cards supported (up from 312)
 - cross-referencing is now more straightforward to new users (much of v1.0 works using the `_ref` option)
   - ``*_ref`` attributes are cross-referenced
      - ``element.nodes`` is not cross-referenced
      - ``element.nodes_ref`` is cross-referenced
 - pickling to reload your deck ~5x faster
 - decreased time required for Case Control Deck with large SETs and many load cases
 - improved optimization checks

OP2:
 - HDF5 export/import support
 - pandas support for matrices
 - couple more results vectorized (e.g., complex strain energy, DMIG strain energy, some forces)
 - grid_point_stressses supported (disabled since v0.7)
 - fixed sparse matrices being stored as dense matrices

GUI:
 - preliminary support for PySide
 - can now mix and match fringe/displacement/vector results (e.g., max principal stress shown on a displaced model)
 - improved animation menu
    - in gui animation
    - more animation profiles
 - bar profile visualzation
 - nominal geometry (useful for deflection plots)
 - improved optimization support
 - improved picking display
 - better PSHELL/PCOMP distinction

Known issues:
 - Transient Pandas Dataframes will fail in newer versions of numpy/pandas.  If anyone knows how to use a MultiIndex,
   this is probably pretty easy to fix.
 
--->
  
# News

### pyNastran v1.0.0 has been released (5/25/2017)
This is a major release.  The focus this time has been on robustness and testing.
Hopefully, it shows.  The software has also been relicensed to be **BSD-3**, which
is a more permissive license and is the same one that numpy, scipy, and
matplotlib use.

Unfortunately, the GUI is more complicated.
 - For open source projects : GPL 2/3
 - For companies that pay a license to Riverbank : proprietary
 - For companies that don't pay a license fee : GPL 2/3

However, you may distribute an unmodified binary.

 - Programmatics:
   - Dropping Python 3.3 support
   - Adding Python 3.6 support

 - [OP2](http://pynastran-git.readthedocs.io/en/latest/quick_start/op2_demo.html)
   - preliminary random results (ATO/CRM/PSD/RMS/NO)
   - improved geometry support
   - MONPNT1/MONPNT3 reading
   - MATPOOL matrices

 - F06 scripts
   - added a preliminary flutter (SOL 145) parser
     - supports multiple subcases
     - PK and PKNL methods supported
     - `plot_Vg_Vf(...)`, `plot_Vg(...)`, `plot_root_locus(...)`
     - input/output units
     - mode switching not fixed yet

 - [GUI](http://pynastran-git.readthedocs.io/en/latest/quick_start/gui.html):
   - complex displacement support
   - animation support
   - vector results (real/complex)
      - SPC Forces, MPC Forces, Load Vector, Applied Load
        - minimal control presently
      - No Grid Point Forces (e.g., freebody loads, interface loads)
   - signficant speedups

 - [BDF](http://pynastran-git.readthedocs.io/en/latest/quick_start/bdf_demo.html)
   - 312 cards supported
   - faster node transforms using:

     ```python
     >>> icd_transform, icp_transform, xyz_cp, nid_cp_cd = get_displacement_index_xyz_cp_cd(dtype='float64, sort_ids=True)
     >>> xyz_cid0 = transform_xyzcp_to_xyz_cid(xyz_cp, icp_transform, cid=0, in_place=False)
     ```

   - simplified card adding
     ```python
     >>> model.add_grid(nid, xyz=[4.,5.,6.], comment='nid, cp, x, y, z')
     ```

- comments can now be created without worrying about `$` signs

     ```python
     >>> model.add_card(['GRID', 10, None, 4.0, 5.0, 6.0], comment='GRID comment\ngrid,nid,cp,x,y,z')
     ```
     ```
     $GRID comment
     $grid,nid,cp,x,y,z
     GRID,10,,4.0,5.0,6.0
     ```

   - unit conversion

- [GUI](http://pynastran-git.readthedocs.io/en/latest/quick_start/gui.html)
   - buttons for picking, rotation center, distance
   - PyQt5 support
   - QScintilla & pygments support for scripting code editor

- [Matlab](http://pynastran-git.readthedocs.io/en/latest/quick_start/matlab.html) integration
   - pyNastran works with Matlab 2014a+

### pyNastran v0.8.0 has been released (8/21/2016)

|  Version  | Docs  | Status |
| :--- 	  | :--- 	  | :--- 	  |
|  [v0.8.0](https://github.com/SteveDoyle2/pyNastran/releases) | [![Documentation Status](https://readthedocs.org/projects/pynastran-git/badge/?version=v0.8)](http://pynastran-git.readthedocs.io/en/v0.8.0/?badge=v0.8.0) | [![Build Status](https://img.shields.io/travis/SteveDoyle2/pyNastran/v0.8.svg)](https://travis-ci.org/SteveDoyle2/pyNastran)  |

<!---

[![Build status](https://ci.appveyor.com/api/projects/status/1qau107h43mbgghi/branch/master?svg=true)](https://ci.appveyor.com/project/SteveDoyle2/pynastran/branch/master)

[![Build status](https://ci.appveyor.com/api/projects/status/1qau107h43mbgghi?svg=true)](https://ci.appveyor.com/project/SteveDoyle2/pynastran)

[![Build Status](https://travis-ci.org/SteveDoyle2/pyNastran.png)](https://travis-ci.org/SteveDoyle2/pyNastran)
[![Requirements Status](https://img.shields.io/requires/github/SteveDoyle2/pyNastran/master.svg)](https://requires.io/github/SteveDoyle2/pyNastran/requirements/?branch=master)	|
--->

See [v0.8.0](https://github.com/SteveDoyle2/pyNastran/releases/tag/v0.8.0) for information regarding enhancements.

### pyNastran v0.7.2 has been Released (4/25/2015)

|  Version  | Docs  |
| :--- 	  | :--- 	  |
|  [v0.7.2](https://github.com/SteveDoyle2/pyNastran/releases) |  [![Documentation Status](https://readthedocs.org/projects/pynastran-git/badge/?version=v0.7.2)](http://pynastran-git.readthedocs.io/en/v0.7.2/?badge=v0.7.2) |

See [v0.7.2](https://github.com/SteveDoyle2/pyNastran/releases) for information regarding enhancements.

### Version 0.6.1 has been released (6/2013)
**Version 0.6** improves BDF reading.  The reader is more robust and also requires proper BDF field formatting (e.g. a integer field can't be a float).  Additionally, cards also have a comment() method.

Marcin Gąsiorek participated in the latest pyNastran under the European Space Agency's (ESA) "Summer of Code In Space" [SOCIS](http://sophia.estec.esa.int/socis2012/?q=node/13) program.  The program provides a stipend to students to work on open-source projects.
He did a great job of simplifying code and creating nicer documentation.


releaseNotes.txt
This document discusses the major features (and limitations) added in each
version of the software.  See the HTML documentation for more information.

If you like it/want to help leave some feedback on the
Dicussion Page:       https://groups.google.com/forum/#!forum/pynastran-discuss
Developer Dicussion:  https://groups.google.com/forum/#!forum/pynastran-dev

If you have a bug/want a new feature or card, leave some feedback on the
Issues List:     https://github.com/SteveDoyle2/pyNastran/issues

-----------------------v1.1.0 released 2018/6/26-----------------------
This is a major release.  As in the v1.0 release, the focus has been on testing.

The bigest change has been to cross-referencing.  All cross-referenced 
objects now have an ``_ref`` attribute that is cross-referened, which 
is ``None`` if the model is not cross referenced.  Objects maintain the
uncross-referenced form as well.  For a CQUAD4, instead of:
 ``model.elements[eid].nodes[0].nid``
the referenced object is:
 ``model.elements[eid].nodes_ref[0].nid``

For detail on the license for the various components, see the v1.0 release.

Quick Overview:
---------------
Programmatics:
 - Added support for numpy 1.14
 - Dropping support for Python 3.4 (2.7, 3.5, 3.6 are supported)
 - Dropping support for VTK 5/6 (7/8 are supported)

BDF:
 - model may be pickled (model.save('model.obj') and model.load('model.obj')
 - simplified cross-referencing

OP2:
 - model may be pickled (model.save('model.obj') and model.load('model.obj')
 - Added support for exporting OP2 to HDF5 (uses pyNastran format, not MSC)
 - real sparse matrices take much less memory now; were being converted to dense matrices

GUI:
 - improved animations
 - improved labels


Detailed Overview:
------------------
BDF:
  - New features:
   - model.get_reduced_loads(load_id)
   - model.get_reduced_dloads(dload_id)
   - card1 == card2 now supported
   - PBARL/PBEAML support the NX TUBE2 type
   - added FREQ3, FREQ5, CAERO5, PAERO5, MATT3, SPLINE3, RSSCON, OMIT1
   - more xref
   - added model.clear_attributes()
   - renumbering:
     - bdf_renumber now supports renumbers SETs and SPLINEx cards
     - SPOINTs/EPOINTs now use dictionaries to enable SPOINT/EPOINT renumbering
     - caero sub-panels ids are now renumbered
     - renumbering mapper object returned now
   - improved removed_unused card support
   - improved mirroring
   - read_bdf StringIO option now parses pyNastran header
   - subcase copying speedup (helps with SETs)
   - preliminary ZONA loading
   - added atmosphere2.make_flfacts_eas_sweep, make_flfacts_mach_sweep, and
     make_flfacts_alt_sweep with an EAS (equivalent airspeed) limiter
   - rotate_v_wa_wb for CBAR/CBEAM to determine 

  - Bug fixes:
   - more add_card documentation (e.g., add_grid, add_ctria3)
   - fixed NSMADD card type (was SPCADD)
   - fixed CPLTSTN3 card type (was CTRIA3)
   - fixed CPLTSTS3 card type (was CTRIA3)
   - fixed shell MCIDs not renumbering
   - fixed FREQx renumbering crash
   - model may be pickled again (model.save('model.obj') and model.load('model.obj')
   - fixed PBARL/PBEAML DBOX error
   - fixed PositionWRT bug
   - fixed LOAD card messing up load ids after cross referencing
   - fixed TRIM default on aeqr (was 0.0/rigid; should be 1.0/elastic) 
   - NLPCI now gets written when there are no other dynamic cards
   - fixed CBUSH cid=0 bug
   - fixed TABLED4/TABLEM4 stopping error

  - API changes:
   - model.Node(nid, allow_empty_nodes=False msg='') no longer supports 
     allow_empty_nodes.  Use:
       model.EmptyNode(nid, msg='') instead for that
       model.Node(nid, msg='') is the new form
   - model.Nodes(nid, allow_empty_nodes=False msg='') no longer supports 
     allow_empty_nodes.  Use:
       model.EmptyNodes(nid, msg='') instead for that
       model.Nodes(nid, msg='') is the new form
   - PCOMPG.validate() now checks that global ply ids are unique
   - xref_nodes_with_elements now creates a list instead of a set
     (fixes a Python 3.x bug)
   - get_MPCx_node_ids_c1 is now get_MPCx_node_ids_c1
     (was inconsistent with what it does)
   - get_MPCx_node_ids_c1 created
   - xref'd objects now use _ref globally
   - aestat.id is now aestat.aestat_id
   - aeparm.id is now aeparm.aeparm_id
   - model.add_aset1/aset (also bset/cset/qset/uset) now a consistent set of
     function arguments and call the same function.  The card will be created
     based on your data instead of necessarily what you asked for.
   - LOAD cards are now stored in model.load_combinations instead of model.loads

  - Known bugs:
   - dynamic loads cross-referencing is buggy; 
     reject the cards if there is a problem
   - PBEAM defaults with ENDA are slightly incorrect.

OP2:
  - New features:
   - added model.set_additional_generalized_tables_to_read(tables) to 
     create custom OP2 readers
   - added complex/average strain energy
   - save/load hdf5 support
   - EIGRL support

  - Bug fixes:
   - improved table skipping
   - fixed RealCShearForceArray f06 writing
   - fixed CEN/3, CEN/4 writing for RealPlateBilinearForceArray
   - improved geometry table reading
   - real sparse matrices take much less memory now; were being converted to dense matrices
   - added RBAR on NX vs. MSC
   - fixed RBE2 with alpha bug
   - fixed CREEP bug
   - fixed RBE3 bug
   - fixed PBCOMP bug

  - API changes:
    - xlsx exporter removed

  - Known bugs:
    - pandas fails on some decks (numpy<1.13 is fine)
    - a large number of PSOLIDs will crash the read_op2/read_op2_geom;
      use PARAM,OGEOM,NO
    - transform_gpforce_to_global doesn't work properly with cylindrical
      or spherical coordinate systems

GUI:
  - New features:
    - control surfaces now get labels (label size doesn't resize properly)
    - in-gui animation
    - delete secondary actor support
    - delete result cases support
    - improved view buttons
    - preferences menu
    - right click support on results sidebar to apply fringe/displacement/veactor results


setup_no_reqs.py
#!/usr/bin/env python
import os
import sys
from setuptools import setup, find_packages

PY2 = False
if sys.version_info < (3, 0):
    PY2 = True
if sys.version_info < (2, 7, 7):
    imajor, minor1, minor2 = sys.version_info[:3]
    # makes sure we don't get the following bug:
    #   Issue #19099: The struct module now supports Unicode format strings.
    sys.exit('Upgrade your Python to >= 2.7.7; version=(%s.%s.%s)' % (imajor, minor1, minor2))

import pyNastran
packages = find_packages()+['gui/icons/*.*']

# set up all icons
icon_path = os.path.join('pyNastran', 'gui', 'icons')
icon_files = os.listdir(icon_path)
icon_files2 = []
for icon_file in icon_files:
    if icon_file.endswith('.png'):
        icon_files2.append(os.path.join(icon_path, icon_file))

exclude_words = []
packages = find_packages(exclude=['ez_setup', 'examples', 'tests'] + exclude_words)
for exclude_word in exclude_words:
    packages = [package for package in packages if exclude_word not in package]
#print(packages, len(packages)) # 83

setup(
    name='pyNastran',
    version=pyNastran.__version__,
    description=pyNastran.__desc__,
    long_description="""\
""",
    classifiers=[
        'Natural Language :: English',
        'Intended Audience :: Science/Research',
        'License :: OSI Approved :: BSD License',
        'Programming Language :: Python :: 2.7',
        'Programming Language :: Python :: 3.4',
        'Programming Language :: Python :: 3.5',
        'Programming Language :: Python :: 3.6',
        ], # Get strings from http://pypi.python.org/pypi?%3Aaction=list_classifiers
    keywords='',
    author=pyNastran.__author__,
    author_email=pyNastran.__email__,
    url=pyNastran.__website__,
    license=pyNastran.__license__,
    packages=packages,
    include_package_data=True,
    zip_safe=False,
    #{'': ['license.txt']}
    #package_data={'': ['*.png']},
    #data_files=[(icon_path, icon_files2)],
    package_data={
        # https://pythonhosted.org/setuptools/setuptools.html#including-data-files
        # If any package contains *.png files, include them:
        '': ['*.png'],
        #'mypkg': ['data/*.dat'],
    },
    entry_points={
        'console_scripts': [
            'test_bdf  = pyNastran.bdf.test.test_bdf:main',
            'test_op2  = pyNastran.op2.test.test_op2:main',
            'test_op4  = pyNastran.op4.test.test_op4:main',
            #'test_abaqus = pyNastran.converters.abaqus.test_abaqus:main',
            'test_pynastrangui = pyNastran.gui.test.test_gui:main',

            'format_converter = pyNastran.converters.type_converter:main',
            'pyNastranGUI = pyNastran.gui.gui:cmd_line',
            'bdf = pyNastran.bdf.mesh_utils.utils:cmd_line',
            'f06 = pyNastran.f06.utils:cmd_line',
        ]
    },
    test_suite='pyNastran.all_tests',
)



pyNastran/__init__.py
import os
import sys

# this variable is automatically set by the .spec file; should be False
is_pynastrangui_exe = False
if is_pynastrangui_exe:
    # pyInstaller
    from pyNastran.version import __version__, __releaseDate__
else:
    import subprocess

    def get_git_revision_short_hash():
        try:
            #ghash = subprocess.check_output(['git', 'rev-parse', '--short', 'HEAD'])

            # independent of pyNastran location as long as there is a git folder
            #   what about if you use setup_user.py install?
            #   what about if you don't have git?
            # can raise a subprocess.CalledProcessError, which means the return code != 0
            ghash = subprocess.check_output(['git', 'describe', '--always'], cwd=os.path.dirname(__file__))

            ghash = ghash.decode('utf-8').rstrip()
        except:
            # git isn't installed
            ghash = 'no.checksum.error'
        return 'dev.%s' % ghash

    revision = get_git_revision_short_hash()
    __version__ = '1.1.0+%s' % revision
    __releaseDate__ = '2018/6/26'
    __releaseDate2__ = 'JUNE 26, 2018'

__author__  = 'Steven Doyle, Michael Redmond, Saullo Castro, hurlei, Paul Blelloch, Nikita Kalutsky'
__email__ = 'mesheb82@gmail.com'
__desc__ = 'Nastran BDF/F06/OP2/OP4 File reader/editor/writer/viewer'
__license__ = 'BSD-3'
__copyright__ = 'Copyright %s; 2011-2018' % __license__
__pyside_copyright__ = 'Copyright LGPLv3 - pySide'
__pyqt_copyright__ = 'Copyright GPLv3 - PyQt'
__website__ = 'https://github.com/SteveDoyle2/pyNastran'
__docs__ = 'http://pynastran-git.readthedocs.io/en/latest/?badge=latest#'
__discussion_forum__ = 'https://groups.google.com/forum/#!forum/pynastran-discuss'

is_release = True  ## turns on skipping of tables that aren't supported


pyNastran/applications/cart3d_nastran_fsi/math_functions.py
from __future__ import print_function
import sys
from six import string_types
from numpy import array, cross, allclose
from numpy.linalg import norm, solve

def is_list_ranged(a, List, b):
    """
    Returns true if a<= x <= b
    or a-x < 0 < b-x
    """
    for x in List:
        if not is_float_ranged(a, x, b):
            return False
    return True

def is_float_ranged(a, x, b):
    """
    Returns true if a<= x <= b
    or a-x < 0 < b-x
    """
    if not a < x:
        if not allclose(x, a):
            return False

    if not x < b:
        if not allclose(x, b):
            return False
    return True


def print_matrix(A):
    msg = ''
    for row in A:
        msg += list_print(row) + '\n'
    return msg

def list_print(listA):
    if len(listA) == 0:
        return '[]'

    msg = '['
    for a in listA:
        if isinstance(a, string_types):
            msg += ' %s,' % (a)
        elif isinstance(a, float):
            msg += ' %-4.2f,' % (a)
        elif isinstance(a, int):
            msg += ' %g,' % (a)
        else:
            try:
                msg += ' %g,' % (a)
            except TypeError:
                print("a = %r" % (a))
                raise

    msg = msg[:-1]
    msg += ' ]'
    return msg


def pierce_plane_vector(p0, p1, p2, pA, pB, pierced_elements):
    """
    http://en.wikipedia.org/wiki/Line-plane_intersection

    [t] = [xa-xb, x1-x0, x2-x0]-1   [xa-x0]
    [u] = [ya-yb, y1-y0, y2-y0]   * [ya-y0]
    [v] = [za-xb, x1-z0, z2-z0]     [za-z0]

    Ax=b          A^-1*b = x
    xa0=mat*tuv   mat*xa0=tuv

    p0,p1,p2 on plane
    a,b are points on the vector
    """
    mat = array([pA-pB, p1-p0, p2-p0]) # numpy array-matrix
    #mat = array([[xa-xb, x1-x0, x2-x0],
    #             [ya-yb, y1-y0, y2-y0],
    #             [za-xb, x1-z0, z2-z0]])
    #xa0 = array([[xa-x0],
    #             [ya-y0],
    #             [za-z0]])
    xa0 = pA-p0  # numpy array
    tuv = solve(mat, xa0)
    return tuv

def distance(x, y):
    """Finds the euclidian/spherical (2D/3D) distance between 2 points"""
    return norm(x - y)


def shepard_SMS_weight(n, nodes):
    """
    a = (R-hi/R*hi)^2
    wi = a/sum(a)

    hi = distance from interpolation point to scatter point
    R  = distance from interpolation point to most distant scatter point

    http://www.ems-i.com/smshelp/Data_Module/Interpolation/Inverse_Distance_Weighted.htm
    Franke & Nielson, 1980
    """
    dists = []
    for nodei in nodes:
        hi = distance(n, nodei)
        dists.append(hi)
    R = min(dists)

    aWeights = []
    for hi in dists:
        a = ((R-hi) / R*hi)**2.
        aWeights.append(a)
    aWeights = array(aWeights)
    weights = aWeights / sum(aWeights)
    return weights


def shepard_weight(n, nodes):
    """
    Finds the weightings based on the distance weighted average method
    In:
        n = node
        nodes = alternate nodes
    Out:
        weights on all nodes (percentage on each node)
    Formula:
        weightI = (1/distanceI)  / sum(1/distanceI) - correct
    http://www.ems-i.com/smshelp/Data_Module/Interpolation/Inverse_Distance_Weighted.htm
    """
    inv_dists = []
    inv_dist_sum = 0.
    d = 0.
    for nodei in nodes:
        hi = distance(n, nodei)
        d += hi
        if allclose(hi, 0.):
            hi = 0.000001  # make this the dominating node

        inv_dist = 1.0 / hi
        inv_dists.append(inv_dist)
        inv_dist_sum += inv_dist

    inv_dists = array(inv_dists)
    weights = inv_dists / inv_dist_sum
    return weights


def area_weight(n, n1, n2, n3):
    """
    Finds the weightings based on the barycentric coordinates weighted average method
    http://www.ems-i.com/smshelp/Data_Module/Interpolation/Inverse_Distance_Weighted.htm
    """
    #a = area(n1, n2, n3)
    a1 = area(a, a2, a3)
    a2 = area(a, a1, a3)
    a3 = area(a, a1, a2)
    a = a1 + a2 + a3
    b1 = a1 / a
    b2 = a2 / a
    b3 = a3 / a

    x = b1 * x1 + b2 * x2 + b3 * x3
    #x = (a1 * x1 + a2 * x2 + a3 * x3) / a
    return w1, w2, w3


def get_triangle_weights(n, n1, n2, n3):
    """
    Returns the weighting for each node such that:
        <F1,F2,F3> = F*<w1,w2,w3>
    on the 3 nodes of a triangle
    """
    weights = shepard_weight(n, [n1, n2, n3])
    w1, w2, w3 = weights
    #w1, w2, w3 = areaWeight(n, n1, n2, n3)
    return w1, w2, w3


def Area(a, b):
    """gets the quad/tri area"""
    return 0.5 * norm(cross(a, b))


def AreaNormal(nodes):
    """
    Returns area,unitNormal
    n = Normal = a x b
    Area   = 1/2 * |a x b|
    V = <v1,v2,v3>
    |V| = sqrt(v1^0.5+v2^0.5+v3^0.5) = norm(V)

    Area = 0.5 * |n|
    unitNormal = n/|n|
    """
    (n0, n1, n2) = nodes
    a = n0 - n1
    b = n0 - n2
    vector = cross(a, b)
    length = norm(vector)
    normal = vector / length
    area = 0.5 * length
    if not allclose(norm(normal), 1.):
        print("a = ", a)
        print("b = ", b)
        print("normal = ", normal)
        print("length = ", length)
        sys.exit('check...')
    return area, normal

def triangle_area_centroid_normal(nodes):
    """Returns area,centroid,unitNormal"""
    (area, normal) = AreaNormal(nodes)
    n1, n2, n3 = nodes[0], nodes[1], nodes[2]
    centroid = Centroid(n1, n2, n3)
    return area, centroid, normal


def Normal(a, b):
    """finds the unit normal vector of 2 vectors"""
    vector = cross(a, b)
    length = norm(vector)
    normal = vector / length
    assert allclose(norm(normal), 1.)
    return normal


def Centroid(A, B, C):
    """returns the centroid of a triangle"""
    centroid = (A + B + C) / 3.
    return centroid



setup_no_gui.py
#!/usr/bin/env python
import os
import sys
from setuptools import setup, find_packages

PY2 = False
if sys.version_info < (3, 0):
    PY2 = True
if sys.version_info < (2, 7, 7):  # 2.7.13 used
    imajor, minor1, minor2 = sys.version_info[:3]
    # makes sure we don't get the following bug:
    #   Issue #19099: The struct module now supports Unicode format strings.
    sys.exit('Upgrade your Python to >= 2.7.7; version=(%s.%s.%s)' % (imajor, minor1, minor2))

import pyNastran
packages = find_packages()+['gui/icons/*.*']
#print "packages = ",packages
#sys.exit()

py2_packages = []
py_packages = []

try:
    import numpy as np
    ver = np.lib.NumpyVersion(np.__version__)
    if ver < '1.11.0':
        print("np.__version__ = %r < '1.11.0'" % np.__version__)
        py_packages.append('numpy >= 1.11.0,<1.13.0')
except ImportError:
    py_packages.append('numpy >= 1.11.0,<1.13.0')

try:
    import scipy
    ver = scipy.version.short_version
    if ver < '0.18.1':
        print("scipy.version.short_version = %r < '0.18.1'" % scipy.version.short_version)
        py_packages.append('scipy >= 0.17.0')
except ImportError:
    py_packages.append('scipy >= 0.18.1')  # 0.18.1 used

try:
    import six
    sver = [int(val) for val in six.__version__.split('-')[0].split('.')]
    if sver < [1, 10, 0]:
        print("six.__version__ = %r < '1.10.0'" % six.__version__)
        py_packages.append('six >= 1.10.0')
except ImportError:
    py_packages.append('six >= 1.10.0')  # 1.10.0 used


#try:
#    import matplotlib
#    sver = [int(val) for val in matplotlib.__version__.split('-')[0].split('.')]
#    if sver < [1, 5, 1]:
#        print("matplotlib.__version__ = %r < '1.5.1'" % six.__version__)
#        py_packages.append('matplotlib >= 1.5.1')
#except ImportError:
#    py_packages.append('matplotlib >= 1.5.1')


try:
    import docopt
    sver = [int(val) for val in docopt.__version__.split('-')[0].split('.')]
    if sver != [0, 6, 2]:
    #if docopt.__version__ != '0.6.2':
        print("docopt.__version__ = %r != '0.6.2'" % docopt.__version__)
        py_packages.append('docopt == 0.6.2')
except ImportError:
    py_packages.append('docopt == 0.6.2')  # 0.6.2 used


try:
    import typing
except ImportError:
    py_packages.append('typing >= 3.6.1')  # 3.6.1 used


if PY2:
    try:
        import pathlib2
    except ImportError:
        py_packages.append('pathlib2 >= 2.2.0')  # 2.2.0 used

    try:
        import scandir
        sver = [int(val) for val in scandir.__version__.split('-')[0].split('.')]
        if sver < [1, 4, 0]:
            print("scandir.__version__ = %r < '1.4.0'" % scandir.__version__)
            py_packages.append('scandir >= 1.4.0')
    except ImportError:
        py_packages.append('scandir >= 1.4.0')  # 1.4.0 used

#py_packages = [
#    'numpy >= 1.9.2',
#    'scipy >= 0.16.0, scipy < 0.18.0',
#]

install_requires = py_packages + [
    # -*- Extra requirements: -*-
    #'docopt == 0.6.2',
    ##'matplotlib >= 1.3.0',
    #'six >= 1.9.0',
    ##'cython',
] + py2_packages,


# set up all icons
icon_path = os.path.join('pyNastran', 'gui', 'icons')
icon_files = os.listdir(icon_path)
icon_files2 = []
for icon_file in icon_files:
    if icon_file.endswith('.png'):
        icon_files2.append(os.path.join(icon_path, icon_file))

exclude_words = []
packages = find_packages(exclude=['ez_setup', 'examples', 'tests'] + exclude_words)
for exclude_word in exclude_words:
    packages = [package for package in packages if exclude_word not in package]
#print(packages, len(packages)) # 83

setup(
    name='pyNastran',
    version=pyNastran.__version__,
    description=pyNastran.__desc__,
    long_description="""\
""",
    classifiers=[
        'Natural Language :: English',
        'Intended Audience :: Science/Research',
        'License :: OSI Approved :: BSD License',
        'Programming Language :: Python :: 2.7',
        'Programming Language :: Python :: 3.4',
        'Programming Language :: Python :: 3.5',
        'Programming Language :: Python :: 3.6',
        ], # Get strings from http://pypi.python.org/pypi?%3Aaction=list_classifiers
    keywords='',
    author=pyNastran.__author__,
    author_email=pyNastran.__email__,
    url=pyNastran.__website__,
    license=pyNastran.__license__,
    packages=packages,
    include_package_data=True,
    zip_safe=False,
    install_requires=install_requires,
    #{'': ['license.txt']}
    #package_data={'': ['*.png']},
    #data_files=[(icon_path, icon_files2)],
    package_data={
        # https://pythonhosted.org/setuptools/setuptools.html#including-data-files
        # If any package contains *.png files, include them:
        '': ['*.png'],
        #'mypkg': ['data/*.dat'],
    },
    entry_points={
        'console_scripts': [
            'run_nastran_double_precision = pyNastran.bdf.test.run_nastran_double_precision:cmd_line',
            'test_bdf  = pyNastran.bdf.test.test_bdf:main',
            'test_op2  = pyNastran.op2.test.test_op2:main',
            'test_op4  = pyNastran.op4.test.test_op4:main',
            'test_abaqus = pyNastran.converters.abaqus.test_abaqus:main',
            'test_pynastrangui = pyNastran.gui.test.test_gui:main',

            'format_converter = pyNastran.converters.type_converter:main',
            'pyNastranGUI = pyNastran.gui.gui:cmd_line',
            'bdf = pyNastran.bdf.mesh_utils.utils:cmd_line',
            'f06 = pyNastran.f06.utils:cmd_line',
        ]# + py2_gui_scripts
    },
    test_suite='pyNastran.all_tests',
)

